CRM/MembershipExtras/Hook/PostProcess/UpdateSubscription.php

```diff
@@ -1,296 +1,331 @@
+
+    foreach ($contributions as $payment) {
+      $installmentCount++;
+      $this->updateContribution(
+        $payment,
+        $formValues['payment_instrument_id'],
+        $installmentCount
+      );
+    }
+  }
+
+  /**
+   * Returns list of contributions associated to current recurring contribution.
+   *
+   * @return array
+   */
+  private function getContributions() {
+    $result = civicrm_api3('Contribution', 'get', [
+      'sequential' => 1,
+      'contribution_recur_id' => $this->recurringContribution['id'],
+      'options' => ['limit' => 0, 'sort' => 'receive_date ASC'],
+    ]);
+
+    if ($result['count'] > 0) {
+      return $result['values'];
+    }
+
+    return array();
+  }
+
+  /**
+   * Calculates first receive date for new cycle day.
+   *
+   * @param string $currentDate
+   * @param int $newCycleDay
+   *
+   * @return string
+   */
+  private function calculateFirstReceiveDate($currentDate, $newCycleDay) {
+    $frequency = $this->recurringContribution['frequency_unit'];
+    $currentCycleDay = CRM_MembershipExtras_Service_CycleDayCalculator::calculate($currentDate, $frequency);
+
+    $difference = $newCycleDay - $currentCycleDay;
+    $absoluteDifference = abs($difference);
+    $interval = "P{$absoluteDifference}D";
+
+    $date = new DateTime($currentDate);
+
+    if ($difference > 0) {
+      $date->add(new DateInterval($interval));
+    } elseif ($difference < 0) {
+      $date->sub(new DateInterval($interval));
+    }
+
+    return $date->format('Y-m-d H:i:s');
+  }
+
+  /**
+   * Respectively updates receive date and/or payment instrument if either of
+   * those were modified for the current recurring contribution.
+   *
+   * @param array $contribution
+   * @param int $instrumentID
+   * @param int $installmentNumber
+   *
+   * @throws \Exception
+   */
+  private function updateContribution($contribution, $instrumentID, $installmentNumber) {
+    if ($contribution['contribution_status'] != 'Pending') {
+      return;
+    }
+
+    $params = [];
+    if ($this->isUpdatedCycleDate() && !$this->isReceiveDateInThePast($contribution)) {
+      $params['receive_date'] = $this->receiveDateCalculator->calculate($installmentNumber);
+
+      if (empty($this->nextContributionDate)) {
+        $this->nextContributionDate = $params['receive_date'];
+      }
+    }
+
+    if ($this->isUpdatedPaymentInstrument()) {
+      $params['payment_instrument_id'] = $instrumentID;
+    }
+
+    if (!empty($params)) {
+      $params['id'] = $contribution['id'];
+      civicrm_api3('Contribution', 'create', $params);
+    }
+  }
+
+  /**
+   * Checks if the given contribution has a receive date in the future.
+   *
+   * @param array $contribution
+   *
+   * @return bool
+   * @throws \Exception
+   */
+  private function isReceiveDateInThePast($contribution) {
+    $now = new DateTime();
+    $receiveDate = new DateTime($contribution['receive_date']);
+
+    if ($receiveDate < $now) {
```

Comment:
you will probably need to rest the time here when comparing or otherwise you might get issues when the date = today date.